packaging/release.py

```diff
@@ -0,0 +1,1410 @@
+
+
+@functools.cache
+def get_remotes() -> Remotes:
+    """Return details about the remotes we need to use."""
+    # assume the devel branch has its upstream remote pointing to the user's fork
+    fork_remote_name = git("branch", "--list", "devel", "--format=%(upstream:remotename)", capture_output=True).stdout.strip()
+
+    if not fork_remote_name:
+        raise ApplicationError("Could not determine the remote for your fork of Ansible.")
+
+    display.show(f"Detected '{fork_remote_name}' as the remote for your fork of Ansible.")
+
+    # assume there is only one ansible org remote, which would allow release testing using another repo in the same org without special configuration
+    all_remotes = git("remote", "-v", capture_output=True).stdout.strip().splitlines()
+    ansible_remote_names = set(line.split()[0] for line in all_remotes if re.search(r"[@/]github[.]com[:/]ansible/", line))
+
+    if not ansible_remote_names:
+        raise ApplicationError(f"Could not determine the remote which '{fork_remote_name}' was forked from.")
+
+    if len(ansible_remote_names) > 1:
+        raise ApplicationError(f"Found multiple candidates for the remote from which '{fork_remote_name}' was forked from: {', '.join(ansible_remote_names)}")
+
+    upstream_remote_name = ansible_remote_names.pop()
+
+    display.show(f"Detected '{upstream_remote_name}' as the remote from which '{fork_remote_name}' was forked from.")
+
+    remotes = Remotes(
+        fork=get_remote(fork_remote_name, push=True),
+        upstream=get_remote(upstream_remote_name, push=False),
+    )
+
+    return remotes
+
+
+def get_upstream_branch(version: Version) -> str:
+    """Return the upstream branch name for the given version."""
+    return f"stable-{version.major}.{version.minor}"
+
+
+def get_git_state(version: Version, allow_stale: bool) -> GitState:
+    """Return information about the current state of the git repository."""
+    remotes = get_remotes()
+    upstream_branch = get_upstream_branch(version)
+
+    git("fetch", remotes.upstream.name, upstream_branch)
+
+    upstream_ref = f"{remotes.upstream.name}/{upstream_branch}"
+    upstream_commit = get_commit(upstream_ref)
+
+    commit = get_commit()
+
+    if commit != upstream_commit:
+        with suppress_when(allow_stale):
+            raise ApplicationError(f"The current commit ({commit}) does not match {upstream_ref} ({upstream_commit}).")
+
+    branch = git("branch", "--show-current", capture_output=True).stdout.strip() or None
+
+    state = GitState(
+        remotes=remotes,
+        branch=branch,
+        commit=commit,
+    )
+
+    return state
+
+
+@functools.cache
+def ensure_venv() -> dict[str, str]:
+    """Ensure the release venv is ready and return the env vars needed to use it."""
+
+    # TODO: consider freezing the ansible and release requirements along with their dependencies
+
+    ansible_requirements = ANSIBLE_REQUIREMENTS_FILE.read_text()
+
+    release_requirements = """
```

Comment:
It was convenient, since all the other requirements are also strings. It didn't seem any more readable when split, and this way avoids having to add a join.